### ISSUE TITLE
"Less than the required minimum number of profiles found" when only specifying one context

### ISSUE THREAD

**Issue by infrascripting** [ansible/ansible#41448]
<!---
Verify first that your issue/request is not already reported on GitHub.
THIS FORM WILL BE READ BY A MACHINE, COMPLETE ALL SECTIONS AS DESCRIBED.
Also test if the latest release, and devel branch are affected too.
ALWAYS add information AFTER (OUTSIDE) these html comments.
Otherwise it may end up being automatically closed by our bot. -->

##### SUMMARY
<!--- Explain the problem briefly -->
When only specifying one `server-side` or only one `client-side` profile, the bigip_virtual_server module fails with: 

```
Less than the required minimum number of profiles found on <path>: Exactly 1 of (UDP Profile (serverside), TCP Profile (serverside), SCTP Profile (serverside))
```

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
<!--- Insert, BELOW THIS COMMENT, the name of the module, plugin, task or feature.
Do not include extra details here, e.g. "vyos_command" not "the network module vyos_command" or the full path-->
bigip_virtual_server

##### ANSIBLE VERSION
<!--- Paste, BELOW THIS COMMENT, verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.4
  config file = /netadm/ansible.cfg
  configured module search path = [u'/home/ansible_netadm/.virtenv/netadm/lib/python2.7/site-packages/ansible/modules', u'/netadm/library']
  ansible python module location = /home/ansible_netadm/.virtenv/netadm/lib/python2.7/site-packages/ansible
  executable location = /home/ansible_netadm/.virtenv/netadm/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

##### CONFIGURATION
<!--- If using Ansible 2.4 or above, paste, BELOW THIS COMMENT, the results of "ansible-config dump --only-changed"
Otherwise, mention any settings you have changed/added/removed in ansible.cfg
(or using the ANSIBLE_* environment variables).-->

```
ANSIBLE_COW_SELECTION(/netadm/ansible.cfg) = random
ANSIBLE_COW_WHITELIST(/netadm/ansible.cfg) = ['bud-frogs', 'bunny', 'cheese', 'daemon', 'default', 'dragon', 'elephant-in-snake', 'elephant', 'eyes', 'hellokitty', 'kitty', 'luke-koala', 'meow', 'milk', 'moofasa', 'moose', 'ren', 'sheep',
ANSIBLE_PIPELINING(/netadm/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/netadm/ansible.cfg) = -o ServerAliveInterval=30 -o ControlMaster=auto -o ControlPersist=4h
ANSIBLE_SSH_CONTROL_PATH(/netadm/ansible.cfg) = /tmp/ansible-ifs-%%h
CACHE_PLUGIN(/netadm/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/netadm/ansible.cfg) = /netadm/.fact_cache
COMMAND_WARNINGS(/netadm/ansible.cfg) = False
DEFAULT_ACTION_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/action']
DEFAULT_BECOME_ASK_PASS(/netadm/ansible.cfg) = False
DEFAULT_BECOME_FLAGS(/netadm/ansible.cfg) = -i
DEFAULT_BECOME_METHOD(/netadm/ansible.cfg) = sudo
DEFAULT_CACHE_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/cache']
DEFAULT_CALLBACK_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/netadm/plugins/callback_plugins', u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/callback']
DEFAULT_CALLBACK_WHITELIST(/netadm/ansible.cfg) = ['timer']
DEFAULT_CONNECTION_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/connection']
DEFAULT_FILTER_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/filter']
DEFAULT_FORKS(/netadm/ansible.cfg) = 200
DEFAULT_GATHERING(/netadm/ansible.cfg) = implicit
DEFAULT_INVENTORY_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/inventory']
DEFAULT_LOAD_CALLBACK_PLUGINS(/netadm/ansible.cfg) = False
DEFAULT_LOOKUP_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/lookup']
DEFAULT_MANAGED_STR(/netadm/ansible.cfg) = Ansible managed: {file} modified on %Y-%m-%d %H:%M by IFS@oclc.org on {host}
DEFAULT_MODULE_LANG(/netadm/ansible.cfg) = C
DEFAULT_MODULE_NAME(/netadm/ansible.cfg) = shell
DEFAULT_MODULE_PATH(/netadm/ansible.cfg) = [u'/home/ansible_netadm/.virtenv/netadm/lib/python2.7/site-packages/ansible/modules', u'/netadm/library']
DEFAULT_POLL_INTERVAL(/netadm/ansible.cfg) = 15
DEFAULT_REMOTE_PORT(/netadm/ansible.cfg) = 22
DEFAULT_REMOTE_USER(/netadm/ansible.cfg) = ansible_netadm
DEFAULT_ROLES_PATH(/netadm/ansible.cfg) = [u'/netadm/roles']
DEFAULT_SQUASH_ACTIONS(/netadm/ansible.cfg) = ['apk', 'apt', 'dnf', 'homebrew', 'pacman', 'pkgng', 'yum', 'zypper', 'increment_serial_numbers_']
DEFAULT_SSH_TRANSFER_METHOD(/netadm/ansible.cfg) = smart
DEFAULT_STDOUT_CALLBACK(/netadm/ansible.cfg) = default
DEFAULT_STRATEGY_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/strategy']
DEFAULT_TIMEOUT(/netadm/ansible.cfg) = 45
DEFAULT_TRANSPORT(/netadm/ansible.cfg) = smart
DEFAULT_VARS_PLUGIN_PATH(/netadm/ansible.cfg) = [u'/opt/ansible/ansible_netadm/.virtualenv/netadm/lib/python2.7/site-packages/ansible/plugins/vars']
DEFAULT_VERBOSITY(/netadm/ansible.cfg) = 1
DEPRECATION_WARNINGS(/netadm/ansible.cfg) = True
HOST_KEY_CHECKING(/netadm/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/netadm/ansible.cfg) = 30
RETRY_FILES_ENABLED(/netadm/ansible.cfg) = False
RETRY_FILES_SAVE_PATH(/netadm/ansible.cfg) = /netadm/inventory
SYSTEM_WARNINGS(/netadm/ansible.cfg) = True
```

##### OS / ENVIRONMENT
<!--- Mention, BELOW THIS COMMENT, the OS you are running Ansible from, and the OS you are
managing, or say "N/A" for anything that is not platform-specific.
Also mention the specific version of what you are trying to control,
e.g. if this is a network bug the version of firmware on the network device.-->
Redhat 7

##### STEPS TO REPRODUCE
<!--- For bugs, show exactly how to reproduce the problem, using a minimal test-case.
For new features, show how the feature would be used. -->

<!--- Paste example playbooks or commands between quotes below -->
```yaml

- name: VIP is created with nodes
  bigip_virtual_server:
      name: "{{ bigip_ltm_vip_name }}"
      destination: "{{ bigip_ltm_prompt_dest_ip_addr }}"
      pool: "{{ bigip_ltm_pool_name }}"
      port: "{{ bigip_ltm_prompt_external_port }}"
      profiles:
        - name: 'tcp-wan-optimized'
          context: 'client-side'
      #  - name: 'tcp-lan-optimized'
      #    context: 'server-side'
      irules: "{{ bigip_ltm_loadbalancers_location }}_{{ bigip_ltm_loadbalancers_environment }}_r_defaultSNAT_001"
      server: "{{ bigip_ltm_active_loadbalancer }}"
      user: "{{ bigip_ltm_admin_user }}"
      password: "{{ bigip_ltm_admin_pass }}"
      validate_certs: "no"
  delegate_to: localhost
  run_once: yes
```

If I uncomment the `tcp-lan-optimized` line, it works perfectly

<!--- You can also paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
<!--- What did you expect to happen when running the steps above? -->
A completed run.

##### ACTUAL RESULTS
<!--- What actually happened? If possible run with extra verbosity (-vvvv) -->
Commenting out the profile with the `server-side` context caused the task to fail.

<!--- Paste verbatim command output between quotes below -->
```
The full traceback is:                                                                                                                                                                                                              [235/1859]
  File "/tmp/ansible_pWt03E/ansible_module_bigip_virtual_server.py", line 1648, in main
    results = mm.exec_module()
  File "/tmp/ansible_pWt03E/ansible_module_bigip_virtual_server.py", line 1406, in exec_module
    raise F5ModuleError(str(e))

fatal: [localhost -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "default_persistence_profile": null,
            "description": null,
            "destination": "redacted",
            "disabled_vlans": null,
            "enabled_vlans": null,
            "fallback_persistence_profile": null,
            "irules": [
                "redacted"
            ],
            "metadata": null,
            "name": "redacted,
            "partition": "Common",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "policies": null,
            "pool": "redacted",
            "port": 443,
            "profiles": [
                {
                    "context": "client-side",
                    "name": "tcp-wan-optimized"
                }
            ],
            "provider": {
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "server": "redacted",
                "server_port": 22,
                "ssh_keyfile": null,
                "timeout": 10,
                "transport": "rest",
                "user": "redacted",
                "validate_certs": false
            },
            "server": "redacted",
            "server_port": 443,
            "snat": null,
            "source": null,
            "state": "present",
            "transport": "rest",
            "user": "redacted",
            "validate_certs": false
        }
    },
    "msg": "400 Unexpected Error: Bad Request for uri: https://redacted:443/mgmt/tm/ltm/virtual/~Common~redacted/\nText: u'{\"code\":400,\"message\":\"01070734:3: Configuration error: Le
ss than the required minimum number of profiles found on /Common/mdc1_shdsvc_v_ifsservers-443_8080: Exactly 1 of (UDP Profile (serverside), TCP Profile (serverside), SCTP Profile (serverside))\",\"errorStack\":[],\"apiError\":3}'"
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/network/f5/bigip_virtual_server.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/f5/bigip_virtual_server.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @caphrim007
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by infrascripting** [comment]
@caphrim007 should I open an issue over at https://github.com/F5Networks/f5-ansible?

**Comment by caphrim007** [comment]
@infrascripting does the same happen in the Ansible stable 2.6 or development code?

No need to open a duplicate issue over there. I'll open it on your behalf as needed. If you want to open future issues there, that's fine

**Comment by infrascripting** [comment]
@caphrim007 Just got the 2.6 stable installed, and it still is giving the same error. I was told that we don't have any existing profiles when we're creating these. I don't know why it's failing.

**Comment by infrascripting** [comment]
@caphrim007 So...I talked to my F5 admin, and he showed me that his setting for the profiles is to set `tcp-lan-optimized` on the client-side, and set the server-side to 'Use Client Profile' in the GUI.

I had him check the config file for me, and it looks like all that does is set up the `tcp-lan-optimized` profile _without_ any type of context. My guess is that by adding context, the requirement to have **some** sort of context applied to both sides was not fulfilled.

I tried setting `tcp-lan-optimized` without any context, and it worked just fine! Now we are able to duplicate what we are setting up in the GUI. Sorry for the false alarm - I misunderstood that we needed a profile that was applicable to both sides.
